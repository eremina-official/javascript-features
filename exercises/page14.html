<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta http-equiv="X-UA-Compatible" content="ie=edge">
  <title>ToggleClass utility</title>
  <style>
      .background-red {
        background-color: red;
      }
  </style>
  <link rel="stylesheet" href="stylesheet.css">
  <script src="page14.js" type="module"></script>

</head>
<body>
  <main>
    <h1>ToggleClass utility</h1>
    <div class="target-div background-red">
      <p>Toggle background color.</p>
    </div>
    <button class="toggle-button">Toggle background color</button>
    <p class="text"></p>
    <p><strong>Practical part:</strong> unit tested in page14.test.js.</p>
  </main>
</body>
</html>