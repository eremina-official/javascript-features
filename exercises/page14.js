console.log('----- toggleClass utility -----');

const targetDiv = document.querySelector('.target-div');
const toggleButton = document.querySelector('.toggle-button');

if (toggleButton) {
  toggleButton.addEventListener('click', toggleBackgroundColor);
}

function toggleBackgroundColor() {
  toggleClass(targetDiv, 'background-red');
}

function toggleClass (nodeName, className) {
  nodeName.classList.toggle(className);
}


console.log('----- working with jsdom -----');

/* 
  const text should be declared inside the changeText() function 
  for the unit test to work. If the const text is declared outside 
  the function jest throws a TypeError: Cannot set property 
  'textContent' of null.
*/
function changeText() {
  const text = document.querySelector('.text');
  text.textContent = 'text';
}


export { toggleClass, changeText };