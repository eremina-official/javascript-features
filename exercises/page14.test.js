import * as page14Module from './page14';
import { toggleClass } from './page14';

describe('toggleClass', () => {
  beforeAll(() => {  
    //set document body HTML
    document.body.innerHTML = 
    '<div class="js-div">' + 
      '<p class ="js-paragraph bg-color-red"></p>' +
    '</div>'; 

    //console.log('beforeAll in describe');
  });

  //console.log('inside describe');

  test('adds class', () => {
    const targetDiv = document.querySelector('.js-div');
    toggleClass(targetDiv, 'bg-color-red');
    expect(targetDiv.classList.contains('bg-color-red')).toBe(true);
    //console.log('inside test');
  });

  test('removes class', () => {
    const targetP = document.querySelector('.js-paragraph'); 
    toggleClass(targetP, 'bg-color-red');
    expect(targetP.classList.contains('bg-color-red')).toBe(false);
  }); 

  afterAll(() => {
    //document.body.innerHTML = '';
  });
});

//console.log('after describe', document.body.innerHTML);


describe('changeText', () => {
  beforeAll(() => {
    document.body.innerHTML = 
    '<div class="text">' + '</div>';
  });

  test('changes inner HTML', () => {
    const text = document.querySelector('.text');
    page14Module.changeText();
    expect(text.textContent).toBe('text');
    expect(text).toMatchSnapshot();
  });
});
