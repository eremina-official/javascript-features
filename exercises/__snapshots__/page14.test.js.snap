// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changeText changes inner HTML 1`] = `
<div
  class="text"
>
  text
</div>
`;
